specifyweb/specify/api.py

```diff
@@ -642,39 +630,102 @@ def handle_to_many(collection, agent, obj, data: Dict[str, Any]) -> None:
-                                     parent_obj=obj)
-            else:
-                # Create a new related object.
-                rel_obj = create_obj(collection, agent, rel_model, rel_data, parent_obj=obj)
-            ids.append(rel_obj.id) # Record the id as one to keep.
-
-        # Delete related objects not in the ids list.
-        # TODO: Check versions for optimistic locking.
-        to_delete = getattr(obj, field_name).exclude(id__in=ids)
-        for rel_obj in to_delete:
-            check_table_permissions(collection, agent, rel_obj, "delete")
-            auditlog.remove(rel_obj, agent, obj)
-        to_delete.delete()
+        if dependent or (isinstance(obj, models.Recordset) and field_name == 'recordsetitems'): 
+            _handle_dependent_to_many(collection, agent, obj, field, val)
+        else: 
+            _handle_independent_to_many(collection, agent, obj, field, val)
+
+def _handle_dependent_to_many(collection, agent, obj, field, value):
+    if not isinstance(value, list): 
+        assert isinstance(value, list), "didn't get inline data for dependent field %s in %s: %r" % (field.name, obj, value)
+        
+    rel_model = field.related_model
+    ids = [] # Ids not in this list will be deleted (if dependent) or removed from obj (if independent) at the end.
+
+    for rel_data in value:
+        rel_data[field.field.name] = obj
+
+        rel_obj = update_or_create_resource(collection, agent, rel_model, rel_data, parent_obj=obj)
+
+        ids.append(rel_obj.id) # Record the id as one to keep.
+
+    # Delete related objects not in the ids list.
+    # TODO: Check versions for optimistic locking.
+    to_remove = getattr(obj, field.name).exclude(id__in=ids).select_for_update()
+    for rel_obj in to_remove:
+        check_table_permissions(collection, agent, rel_obj, "delete")
+        auditlog.remove(rel_obj, agent, obj)
+    
+    to_remove.delete()
+
+class IndependentInline(TypedDict): 
+    update: List[Union[str, Dict[str, Any]]]
+    remove: List[str]
+
+def _handle_independent_to_many(collection, agent, obj, field, value: IndependentInline): 
+    logger.warning("Updating independent collections via the API is experimental and the structure may be changed in the future")
+    
+    rel_model = field.related_model
+
+    to_update = value.get('update', [])
+    to_remove = value.get('remove', [])
+
+    ids_to_fetch = []
+    cached_objs = dict()
+    fk_model = None
+
+    to_fetch = [*to_update, *to_remove]
+
+    # Fetch the related records which are provided as strings
+    for rel_data in to_fetch: 
+        if not isinstance(rel_data, str): continue
+        fk_model, fk_id = strict_uri_to_model(rel_data, rel_model.__name__)
+        ids_to_fetch.append(fk_id)
+
+    if fk_model is not None: 
+        cached_objs = {item.id: obj_to_data(item) for item in get_model(fk_model).objects.filter(id__in=ids_to_fetch).select_for_update()}
```

Comment:
Nice attempt! Note that this can be simplified by simply excluding the ones which reverse side as the current obj using an exclude(). In fact, you won’t even need lines 701-705 in that case. Plus, this is an optimization, on a lot of levels.